mpf/devices/segment_display/segment_display.py

```diff
@@ -251,31 +261,93 @@ def _update_stack(self) -> None:
 
             # update colors if specified
             if top_text_stack_entry.colors:
-                self.colors = top_text_stack_entry.colors
+                colors = top_text_stack_entry.colors
+            else:
+                colors = self._current_state.colors
 
             # update the display
-            self._update_display(new_text, top_text_stack_entry.colors)
+            self._update_display(SegmentDisplayState(new_text, colors, flashing, flash_mask))
 
     def _current_placeholder_changed(self, *args, **kwargs) -> None:
         """Update display when a placeholder changes (callback function)."""
         del args
         del kwargs
         new_text, future = self._current_placeholder.evaluate_and_subscribe({})
         future.add_done_callback(self._current_placeholder_changed)
-        self._update_display(new_text)
+        self._update_display(SegmentDisplayState(new_text, self._current_state.colors, self._current_state.flashing,
+                                                 self._current_state.flash_mask))
 
-    def _update_display(self, new_text: str, new_colors: Optional[List[RGBColor]] = None) -> None:
+    def set_flashing(self, flashing: FlashingType, flash_mask: str = ""):
+        """Enable/Disable flashing."""
+        self._update_display(SegmentDisplayState(self._current_state.text, self._current_state.colors, flashing,
+                                                 flash_mask))
+
+    def set_color(self, colors: List[RGBColor]):
+        """Set display colors."""
+        assert isinstance(colors, list)
+        assert self.hw_display is not None
+        self._update_display(SegmentDisplayState(self._current_state.text,
+                                                 self._expand_colors(colors, len(self._current_state.text)),
+                                                 self._current_state.flashing,
+                                                 self._current_state.flash_mask))
+
+    @property
+    def text(self):
+        """Return current text."""
+        return self._current_state.text
+
+    @property
+    def colors(self):
+        """Return current colors."""
+        return self._current_state.colors
+
+    @property
+    def flashing(self):
+        """Return current flashing state."""
+        return self._current_state.flashing
+
+    @property
+    def flash_mask(self):
+        """Return current flash mask."""
+        return self._current_state.flash_mask
+
+    def _update_display(self, new_state: SegmentDisplayState) -> None:
         """Update display to current text."""
         assert self.hw_display is not None
+        if new_state == self._current_state:
+            return
+
+        self._current_state = new_state
+        text = new_state.text
+        colors = new_state.colors
+        flashing = new_state.flashing
+        flash_mask = new_state.flash_mask
+
+        # make sure text is the same length as the display
+        if len(text) > self.size:
```

Comment:
I don't this this will work when using integrated dots/periods or commas as they are collapsed when displaying them shortening the actual length of the text. For example, with displays that support integrated commas, the following text is legal to be sent to a display of size 7: "1,000,000" even though len(text) is 9 in this case.

Reply:
In the text transition calculations, I created lists of characters that each had additional attributes (color, integrated dot, integrated comma). It was the easiest way to perform operations on the text while preserving colors and integrated dots/commas. Perhaps we might want to do something similar here or even create a SegmentDisplayCharacter class to use throughout all segment display-related classes (including the platforms).

Reply:
Makes sense. Thanks! We definitely need a test for that.